### PR TITLE
CBL-7976 Implement Parse and Compare Non-numeric checkpoint resolution

### PR DESCRIPTION
**Pull request by nitishgupta25**
Implementation is done to enhance the checkpoint resolution process to be able to process and compare non-numeric checkpoints so that the replication can continue from the lower checkpoint if they are mismatched, instead of starting from the beginning.

Jira ticket : https://jira.issues.couchbase.com/browse/CBL-7975

Design doc for change: https://docs.google.com/document/d/1fIQsdT8Xkhs4HcH94POYV2mBlNA4gSHczMd3x5Qi69E/edit?tab=t.0

### CODE REVIEW

File: Replicator/ParsedSequenceID.hh [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        static bool parse(const std::string& str, ParsedSequenceID& out) {
+            if ( str.empty() ) return false;
+
+            // Split into up to 4 tokens by ':' (4th would mean invalid).
+            uint64_t values[3] = {};
+            bool     empty[3]  = {};
+            int      count     = 0;
+            size_t   start     = 0;
+
+            for ( size_t i = 0; i <= str.size(); ++i ) {
```
Comment (jianminzhao) [inline Replicator/ParsedSequenceID.hh]: you may use litecore::split in StringUtil.hh to split by ':', and you can use slice_istream parse a uint64_t

File: Replicator/tests/ParsedSequenceIDTest.cc [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+    }
+
+    SECTION("both components equal") {
+        auto s = mustParse("50:50");
+        CHECK(s.triggeredBy() == 50);
+        CHECK(s.seq() == 50);
+        CHECK(s.isBackfillFormat());
+    }
+
+    SECTION("triggeredBy larger than seq") {
```
Comment (adamcfraser) [inline Replicator/tests/ParsedSequenceIDTest.cc]: This looks like the same case as 'basic backfill' (triggeredBy larger than Seq).  You shouldn't need to handle the reverse case (200:300 is an invalid triggeredBy/Seq combination).

File: Replicator/ParsedSequenceID.hh [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        ParsedSequenceID() = default;
+
+        ParsedSequenceID(uint64_t seq, uint64_t triggeredBy, uint64_t lowSeq) : _seq(seq), _triggeredBy(triggeredBy), _lowSeq(lowSeq) {}
+
+        [[nodiscard]] uint64_t seq()         const { return _seq; }
+        [[nodiscard]] uint64_t triggeredBy() const { return _triggeredBy; }
+        [[nodiscard]] uint64_t lowSeq()      const { return _lowSeq; }
+
+        [[nodiscard]] bool isLowSeqFormat()    const { return _lowSeq > 0; }
+        [[nodiscard]] bool isBackfillFormat()  const { return _lowSeq == 0 && _triggeredBy > 0; }
```
Comment (adamcfraser) [inline Replicator/ParsedSequenceID.hh]: I don't think 'isBackfillFormat' is exactly the right description for this, as lowSeq:triggeredBy:seq **does** represent a backfill - it's just one that's happening while lowSeq is also populated.

File: Replicator/tests/ParsedSequenceIDTest.cc [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+    }
+
+    SECTION("same triggeredBy, same seq → not before") {
+        CHECK_FALSE(mustParse("50:50").before(mustParse("50:50")));
+    }
+}
+
+TEST_CASE("ParsedSequenceID before - LowSeq vs LowSeq", "[ParsedSequenceID]") {
+    // Both "LowSeq:TB:Seq": compare lowSeq first, then inner pair as tie-break.
+
```
Comment (adamcfraser) [inline Replicator/tests/ParsedSequenceIDTest.cc:205]: This section needs a case confirming "20:100:35" is before "20::100".  (the backfill boundary case, but with matching lowseq).

**Review by adamcfraser — DISMISSED** [review]
Generally looks good, some comments and recommendations on an additional test case.

File: Replicator/tests/ParsedSequenceIDTest.cc (hunk 2 above)
Comment (nitishgupta25) [reply Replicator/tests/ParsedSequenceIDTest.cc]: Agreed — "triggeredBy larger than seq" with 200:1 is structurally identical to "basic backfill". Since TriggeredBy < Seq (e.g. 200:300) is an invalid SG combination, Will remove the section.

File: Replicator/tests/ParsedSequenceIDTest.cc (hunk 4 above)
Comment (nitishgupta25) [reply Replicator/tests/ParsedSequenceIDTest.cc:205]: will add a section to testing "20:100:35" before "20::100". The logic already handles it correctly (tie-break recurses into inner backfill-vs-simple comparison, 100 <= 100 → true), but the test was missing. Will also add the reverse check (false, due to strict <) to confirm the asymmetry.

File: Replicator/ParsedSequenceID.hh (hunk 3 above)
Comment (nitishgupta25) [reply Replicator/ParsedSequenceID.hh]: yes..got it. renamed to isTriggeredRevision() to describe the format structurally - it identifies the "TriggeredBy:Seq" format where triggeredBy is the primary comparison value and no lowSeq is present. 
This avoids implying that LowSeq:TriggeredBy:Seq is not a backfill. Similarly renamed isLowSeqFormat() to isLowSeqRevision() to keep all three classifiers consistent with the existing isSimpleRevision() naming pattern:
isSimpleRevision() — "Seq"
isTriggeredRevision() — "TriggeredBy:Seq"
isLowSeqRevision() — "LowSeq:TriggeredBy:Seq" / "LowSeq::Seq"

File: Replicator/ParsedSequenceID.hh (hunk 1 above)
Comment (nitishgupta25) [reply Replicator/ParsedSequenceID.hh]: I think yes using  litecore::split(str, ":") to split the components and a readUInt lambda backed by slice_istream::readDecimal() + eof() to parse each uint64_t. This removes the manual for loop, the std::stoull exception handling, and the explicit negative-value guard..thanks

**Review by adamcfraser — COMMENTED** [review]
LGTM from a functional perspective - test cases are comprehensive and sequence comparison algorithm is correct.  Leaving to Jianmin for final approval of the implementation details.

File: Replicator/ParsedSequenceID.hh [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                    if ( !readUInt(parts[0], v[0]) ) return false;
+                    if ( !parts[1].empty() && !readUInt(parts[1], v[1]) ) return false;  // empty → 0 for "LowSeq::Seq"
+                    if ( !readUInt(parts[2], v[2]) ) return false;
+                    out = ParsedSequenceID(v[2], v[1], v[0]);  // LowSeq:TriggeredBy:Seq
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
```
Comment (jianminzhao) [inline Replicator/ParsedSequenceID.hh:107]: I am using https://github.com/couchbase/sync_gateway/blob/main/db/sequence_id.go#L187 as reference for "truth."

File: Replicator/ParsedSequenceID.hh [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        [[nodiscard]] bool before(const ParsedSequenceID& seqID2) const {
+            const auto& a = *this;
+            const auto& b = seqID2;
+
+            if ( a.isLowSeqRevision() ) {
+                if ( b.isLowSeqRevision() ) {
+                    if ( a.lowSeq() != b.lowSeq() ) return a.lowSeq() < b.lowSeq();
```
Comment (jianminzhao) [inline Replicator/ParsedSequenceID.hh:114]: Per the reference, n1:100:m1 .before. n2:101:m2  for any n and m, but the this rule compares n1 and n2 they are both greater than 0.

**Review by jianminzhao — CHANGES_REQUESTED** [review]

File: Replicator/ParsedSequenceID.hh (hunk 5 above)
Comment (nitishgupta25) [reply Replicator/ParsedSequenceID.hh:107]: We should not use the safesequence comparison...instead you should use this corrected code which is recently done by adam:
https://github.com/couchbase/sync_gateway/commit/0bc83f2ad2428748dd5c295897b78ffc18e52331

File: Replicator/ParsedSequenceID.hh (hunk 6 above)
Comment (nitishgupta25) [reply Replicator/ParsedSequenceID.hh:114]: yes this is expected..

File: Replicator/Checkpoint.cc [hunk 7]
```diff
@@ -114,18 +114,20 @@ namespace litecore::repl {
...
                     match   = false;
                 } else {
-                    LogTo(SyncLog, "Ignoring remote sequence on server since client side is older, some redundant "
-                                   "changes may be proposed...");
+                    LogTo(SyncLog, "Ignoring remote sequence on server since client side is older or equal, some "
+                                   "redundant changes may be proposed...");
                 }
             } else {
-                Warn("Non-numeric remote sequence detected, resetting replication back to start.  Redundant changes "
+                Warn("Unparseable remote sequence detected, resetting replication back to start.  Redundant changes "
```
Comment (jianminzhao) [inline Replicator/Checkpoint.cc]: whether _remote is unparseable or remoteSeuqnece._remote is unparseable may be useful when debugging. In theory, _remote should be parseable because it is early saved, but explicit noting which is not parseable may avoid "guessing" at debugging CBSE.

**Review by jianminzhao — CHANGES_REQUESTED** [review]

File: Replicator/Checkpoint.cc (hunk 7 above)
Comment (nitishgupta25) [reply Replicator/Checkpoint.cc]: yes agree...i will correct this ambiguity

**Review by jianminzhao — APPROVED** [review]

**Review by jianminzhao — APPROVED** [review]
